src/lib/Libifl/pbsD_deljoblist.c

```diff
@@ -46,160 +46,554 @@
  */
-int
-PBSD_deljoblist_put(int c, int function, char **jobids, int numofjobs, char *extend, int prot, char **msgid)
+static struct batch_deljob_status *
+PBSD_deljoblist(int c, int function, char **jobids, int numjids, char *extend)
 {
-	int rc;
+	int i;
+	svr_jobid_list_t *svr_itr = NULL;
+	struct batch_deljob_status *retlist = NULL;
+	svr_jobid_list_t *svr_jobid_list_hd = NULL;
+	char **bcastjids = NULL;
+	int bcastjidssize = 0;
+	int nbcastids = 0;
+	int fd_virtual = is_fd_virtual(c);
 
-	if (prot == PROT_TCP) {
-		DIS_tcp_funcs();
-	} else {
-		if ((rc = is_compose_cmd(c, IS_CMD, msgid)) != DIS_SUCCESS)
-			return rc;
-	}
+	if (fd_virtual) {	/* Virtual fd means we can talk to more than 1 server instance */
+		/* Club jobs by each server */
+		for (i = 0; i < numjids; i++) {
+			int fd;
+
+			fd = get_server_fd_from_jid(c, jobids[i]);
 
-	if ((rc = encode_DIS_ReqHdr(c, function, pbs_current_user)) ||
-		(rc = encode_DIS_JobsList(c, jobids, numofjobs)) ||
-		(rc = encode_DIS_ReqExtend(c, extend))) {
-		if (prot == PROT_TCP) {
-			if (set_conn_errtxt(c, dis_emsg[rc]) != 0)
-				return (pbs_errno = PBSE_SYSTEM);
+			if (fd == -1) {
+				if (add_failed_job(jobids[i], PBSE_UNKJOBID, &retlist) != 0)
+					goto err;
+			} else {
+				if (add_jid_to_list_by_fd(jobids[i], fd, &svr_jobid_list_hd) != 0)
+					goto err;
+			}
 		}
-		return (pbs_errno = PBSE_PROTOCOL);
+	} else {	/* Single server/real fd to a specific server instance */
+		svr_jobid_list_hd = malloc(sizeof(svr_jobid_list_t));
+		if (svr_jobid_list_hd == NULL) {
+			pbs_errno = PBSE_SYSTEM;
+			goto err;
+		}
+		svr_jobid_list_hd->total_jobs = numjids;
+		svr_jobid_list_hd->svr_fd = c;
+		svr_jobid_list_hd->jobids = jobids;
+		svr_jobid_list_hd->next = NULL;
 	}
 
-	if (prot == PROT_TPP) {
-		pbs_errno = PBSE_NONE;
-		if (dis_flush(c))
-			pbs_errno = PBSE_PROTOCOL;
-		return pbs_errno;
+	/* initialize the thread context data, if not initialized */
+	if (pbs_client_thread_init_thread_context() != 0)
+		goto err;
+
+	/* For each server, get the list of jobs to delete and send the request over */
+	for (svr_itr = svr_jobid_list_hd; svr_itr != NULL; svr_itr = svr_itr->next) {
+		if (pbs_client_thread_lock_connection(svr_itr->svr_fd) != 0)
+			goto err;
+		for (i = 0; i < svr_itr->total_jobs; i = i + numjids) {
+			struct batch_deljob_status *ret;
+
+			numjids = NUM_JOBS_IN_A_BATCH;
+			numjids = ((i + numjids) <= svr_itr->total_jobs) ? numjids : (svr_itr->total_jobs - i);
+
+			ret = deljob_to_server(svr_itr->svr_fd, function, &(svr_itr->jobids[i]), numjids, extend);
+
+			if (ret != NULL) {
+				struct batch_deljob_status *iter_list;
+				struct batch_deljob_status *prev = NULL;
+				struct batch_deljob_status *next = NULL;
+
+				for (iter_list = ret; iter_list != NULL; prev = iter_list, iter_list = next) {
+					next = iter_list->next;
+					if (iter_list->code == PBSE_UNKJOBID && fd_virtual) {
+						/* Add job to the broadcast list */
+						if (nbcastids >= (bcastjidssize - 1)) {
```

Comment:
Can add jobs to broadcast list be a separate function. You can hide all the buffer reallocation etc inside the function. It will also make PBSD_deljoblist() more readable. Its 100+ lines now.